### PR TITLE
cmd/snap: stacktraces debug endpoint

### PR DESCRIPTION
**Pull request by stolowski**
Sometimes (such as when someone reports download getting stuck inside download handler) it might be useful to get a stacktrace while snapd is still running (and not require `kill -sigabrt ...` to obtain it from go), so I'm proposing a simple debug stacktrace command. For security this debug is restricted to root only.

Example:
```
$ sudo ./snap debug stacktrace
goroutine 328 [running]:
github.com/snapcore/snapd/daemon.getStacktrace(0xc0006f4375, 0xa)
	/home/vagrant/go/src/github.com/snapcore/snapd/daemon/api_debug_stacktrace.go:30 +0x6d
github.com/snapcore/snapd/daemon.getDebug(0x1864760, 0xc00011e300, 0x0, 0x0, 0x0)
	/home/vagrant/go/src/github.com/snapcore/snapd/daemon/api_debug.go:294 +0x231
github.com/snapcore/snapd/daemon.(*Command).ServeHTTP(0x1864760, 0x11b4520, 0xc0004820a0, 0xc00011e200)
	/home/vagrant/go/src/github.com/snapcore/snapd/daemon/daemon.go:232 +0x54a
github.com/snapcore/snapd/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc0002c2300, 0x11b4520, 0xc0004820a0, 0xc00011e000)
	/home/vagrant/go/src/github.com/snapcore/snapd/vendor/github.com/gorilla/mux/mux.go:212 +0xe2
github.com/snapcore/snapd/daemon.logit.func1(0x11b4d60, 0xc0002b2000, 0xc00011e000)
	/home/vagrant/go/src/github.com/snapcore/snapd/daemon/daemon.go:280 +0xd8
net/http.HandlerFunc.ServeHTTP(0xc00049c500, 0x11b4d60, 0xc0002b2000, 0xc00011e000)
	/usr/lib/go-1.13/src/net/http/server.go:2007 +0x44
net/http.serverHandler.ServeHTTP(0xc0000ac700, 0x11b4d60, 0xc0002b2000, 0xc00011e000)
	/usr/lib/go-1.13/src/net/http/server.go:2802 +0xa4
net/http.(*conn).serve(0xc0004fc3c0, 0x11ba160, 0xc000078000)
	/usr/lib/go-1.13/src/net/http/server.go:1890 +0x875
created by net/http.(*Server).Serve
	/usr/lib/go-1.13/src/net/http/server.go:2928 +0x384

goroutine 1 [select, 4 minutes]:
main.run(0xc0003e9140, 0x0, 0x0)
	/home/vagrant/go/src/github.com/snapcore/snapd/cmd/snapd/main.go:153 +0x4e3
main.main()
	/home/vagrant/go/src/github.com/snapcore/snapd/cmd/snapd/main.go:64 +0x10c

goroutine 19 [syscall, 4 minutes]:
os/signal.signal_recv(0x0)
	/usr/lib/go-1.13/src/runtime/sigqueue.go:147 +0x9c
os/signal.loop()
	/usr/lib/go-1.13/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/lib/go-1.13/src/os/signal/signal_unix.go:29 +0x41

goroutine 266 [IO wait]:
internal/poll.runtime_pollWait(0x7fa2680d8af8, 0x72, 0xffffffffffffffff)
	/usr/lib/go-1.13/src/runtime/netpoll.go:184 +0x55
internal/poll.(*pollDesc).wait(0xc000105318, 0x72, 0xe00, 0xe4c, 0xffffffffffffffff)
	/usr/lib/go-1.13/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	/usr/lib/go-1.13/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc000105300, 0xc000349000, 0xe4c, 0xe4c, 0x0, 0x0, 0x0)
	/usr/lib/go-1.13/src/internal/poll/fd_unix.go:169 +0x1cf
net.(*netFD).Read(0xc000105300, 0xc000349000, 0xe4c, 0xe4c, 0x203000, 0x0, 0xe3f)
	/usr/lib/go-1.13/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc00014c020, 0xc000349000, 0xe4c, 0xe4c, 0x0, 0x0, 0x0)
	/usr/lib/go-1.13/src/net/net.go:184 +0x68
crypto/tls.(*atLeastReader).Read(0xc00044b900, 0xc000349000, 0xe4c, 0xe4c, 0x3, 0x8, 0xc0004d68a0)
	/usr/lib/go-1.13/src/crypto/tls/conn.go:780 +0x60
bytes.(*Buffer).ReadFrom(0xc0004dc958, 0x119cc60, 0xc00044b900, 0x40c025, 0xe6e040, 0xf5b940)
	/usr/lib/go-1.13/src/bytes/buffer.go:204 +0xb4
crypto/tls.(*Conn).readFromUntil(0xc0004dc700, 0x119e840, 0xc00014c020, 0x5, 0xc00014c020, 0x40e046)
	/usr/lib/go-1.13/src/crypto/tls/conn.go:802 +0xec
crypto/tls.(*Conn).readRecordOrCCS(0xc0004dc700, 0x0, 0x0, 0x3)
	/usr/lib/go-1.13/src/crypto/tls/conn.go:609 +0x124
crypto/tls.(*Conn).readRecord(...)
	/usr/lib/go-1.13/src/crypto/tls/conn.go:577
crypto/tls.(*Conn).Read(0xc0004dc700, 0xc000558000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/go-1.13/src/crypto/tls/conn.go:1255 +0x161
net/http.(*persistConn).Read(0xc000369b00, 0xc000558000, 0x1000, 0x1000, 0xc0004b05a0, 0xc0004d6c20, 0x406af5)
	/usr/lib/go-1.13/src/net/http/transport.go:1758 +0x75
bufio.(*Reader).fill(0xc000722c60)
	/usr/lib/go-1.13/src/bufio/bufio.go:100 +0x103
bufio.(*Reader).Peek(0xc000722c60, 0x1, 0x0, 0x0, 0x1, 0xc0004b0400, 0x0)
	/usr/lib/go-1.13/src/bufio/bufio.go:138 +0x4f
net/http.(*persistConn).readLoop(0xc000369b00)
	/usr/lib/go-1.13/src/net/http/transport.go:1911 +0x1d6
created by net/http.(*Transport).dialConn
	/usr/lib/go-1.13/src/net/http/transport.go:1580 +0xb0d

goroutine 354 [syscall]:
.....
```

### CODE REVIEW

File: daemon/api_debug_stacktrace.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"runtime"
+	"strings"
+)
+
+func getStacktrace() Response {
+	stack := make([]byte, 65535)
+	all := true
+	n := runtime.Stack(stack, all)
+	data := strings.Split(string(stack[:n]), "\n")
```
Comment (bboozzoo) [inline daemon/api_debug_stacktrace.go:31]: Why split the string into lines?

File: daemon/api_debug.go [hunk 2]
```diff
@@ -36,10 +36,10 @@ import (
 )
 
 var debugCmd = &Command{
-	Path:   "/v2/debug",
-	UserOK: true,
-	GET:    getDebug,
-	POST:   postDebug,
+	Path:     "/v2/debug",
+	RootOnly: true,
```
Comment (bboozzoo) [inline daemon/api_debug.go:40]: I think this should remain UserOK. For pprof there's a separate debug command with PathPrefix set to /v2/debug/pprof added right before the actual debugCmd. I think you can try and register a new api.Command with Path set to /v2/debug/stacktrace and RootOnly, but add it right before debugCmd.

File: daemon/api_debug_stacktrace.go (hunk 1 above)
Comment (stolowski) [reply daemon/api_debug_stacktrace.go:31]: No deep reason / no strong opinion. I'm fine with just returning a text blob.

File: daemon/api_debug.go (hunk 2 above)
Comment (stolowski) [reply daemon/api_debug.go:40]: Done

File: daemon/api_debug_stacktrace.go (hunk 1 above)
Comment (stolowski) [reply daemon/api_debug_stacktrace.go]: Done.

**Review by mardy — APPROVED** [review]

**Comment by pedronis** [comment]
I'm marking it blocked because we are about to merge the new way to do access in daemon, and I don't want to delay that further, this will need to get a master merge and adapt

**Review by mvo5 — COMMENTED** [review]
This looks great, thank you! Maybe add a small (spread) test to demo that it works?

**Review by pedronis — COMMENTED** [review]
comment about the name

File: tests/main/snap-debug-stacktrace/task.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+summary: Ensure `snap debug stacktrace` command works
+
+execute: |
+  snap debug stacktrace | MATCH 'goroutine'
+  snap debug stacktrace | MATCH 'github.com/snapcore/snapd/daemon'
+
+  echo "Ensure that stacktrace is not available to non-root users"
```
[the remark this answers is not in the document]
Comment (stolowski) [reply tests/main/snap-debug-stacktrace/task.yaml:7]: Yeah, that is now the case with the new changes. I suppose the test will fail now until tightened up.

[the remark this answers is not in the document]
Comment (mvo5) [reply tests/main/snap-debug-stacktrace/task.yaml:7]: Good point, I wonder if we should limit it to root initially?

**Review by mvo5 — APPROVED** [review]
Very nice, thank you! Approving but we need to thing if anyone or only root should be able to trigger this.

File: tests/main/snap-debug-stacktrace/task.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (stolowski) [reply tests/main/snap-debug-stacktrace/task.yaml:7]: This should be good now with changes from Samuele in master.

**Review by bboozzoo — APPROVED** [review]
LGTM

**Comment by pedronis** [comment]
the name of the command and the endpoint are not consistent

**Pull request merged**